src/IECoreScene/MeshAlgoTangents.cpp

```diff
@@ -239,3 +261,221 @@ std::pair<PrimitiveVariable, PrimitiveVariable> IECoreScene::MeshAlgo::calculate
 
 	return std::make_pair( tangentPrimVar, bitangentPrimVar );
 }
+
+std::pair<PrimitiveVariable, PrimitiveVariable> IECoreScene::MeshAlgo::calculateTangentsFromPrimitiveCentroid(
+	const MeshPrimitive *mesh,
+	const std::string &position /* = "P" */,
+	const std::string &normal /* = "N" */,
+	bool orthoTangents,
+	bool leftHanded
+)
+{
+	// get point and normal data
+	const V3fVectorData *positionData = mesh->variableData<V3fVectorData>( position );
+	const V3fVectorData *normalData = mesh->variableData<V3fVectorData>( normal );
+	if( !positionData )
+	{
+		std::string e = boost::str( boost::format( "MeshAlgo::calculateTangentsFromPrimitiveCentroid : MeshPrimitive has no Vertex \"%s\" primitive variable." ) % position );
+		throw InvalidArgumentException( e );
+	}
+	if( !normalData )
+	{
+		std::string e = boost::str( boost::format( "MeshAlgo::calculateTangentsFromPrimitiveCentroid : MeshPrimitive has no Vertex \"%s\" primitive variable." ) % normal );
+		throw InvalidArgumentException( e );
+	}
+	const V3fVectorData::ValueType &points = positionData->readable();
+	const V3fVectorData::ValueType &normals = normalData->readable();
+
+	int numPoints = points.size();
+	std::vector<V3f> uTangents( numPoints, V3f( 0 ) );
+	std::vector<V3f> vTangents( numPoints, V3f( 0 ) );
+
+	const IntVectorData *vertsPerFaceData = mesh->verticesPerFace();
+	const IntVectorData::ValueType &vertsPerFace = vertsPerFaceData->readable();
+
+	const IntVectorData *vertIdsData = mesh->vertexIds();
+	const IntVectorData::ValueType &vertIds = vertIdsData->readable();
+
+	std::vector<V3f> centroids( vertsPerFace.size(), V3f( 0 ) );
+	std::vector<int> faceIdPerVert( vertIds.size(), -1 );
+
+	// calculate centroids
```

Comment:
I'm surprised we don't already have centroid available via the `PrimitiveEvaluator`... we have `centerOfGravity()` which seems considerably harder to compute... Maybe just add a todo for now to port this there (or to `MeshAlgo::calculateCentroids` perhaps).